### PR TITLE
Load classes from opaque calls [TG-3657] 

### PR DESCRIPTION
**Pull request by thk123**
Based on #2302 :white_check_mark: 

When we come to an opaque method, we need to load the return type (And I assume the parameters). I think this resolves the issue @smowton had with #2267 (Specifically, adding a method that _wasn't_ populated to the populated list, that then forced (I think) symex driven loading to load the method. This, apart from being counter intuitive, meant that if the `cproverNondetInitialize` called any other methods, these would not be loaded. 

Further I think we should be loading the relevant classes of returns of opaque methods. The parameters I am less sure about. 

This needs some tests to verify it is working as expected and of course a TG bump. I also will probably turn the lambda into a real method.

### CODE REVIEW

**Comment by smowton** [comment]
This doesn't relate to symex-driven loading -- `ci_lazy_methods.cpp` et al is old-school lazy-loading-v1. I'm not sure how the method bodies were getting populated, but indeed it was surely wrong that it was added to the already-done list without being done.

Re: parameters -- effectively here ci-lazy-methods is mimicking the behaviour of java-opaque-method-stubs et al, and doing it inexactly (different users of the Java frontend will choose different stub strategies, including "always return zero," in which case this will be overkill). However we can safely approximate that a stub might be a source of new instances of its return type, and any type that return type can carry references to.

It can't be a source of instances of its parameter types, since we don't have out parameters in Java, but it can be a source of types they point to if we have mocks that alter passed objects. It so happens we know no commonly-used stubbing strategy actually does that though, so it's up to you whether you do it. 

Along similar lines, if a stub returns an `A { Set s };`, then should we assume it's a source of `HashSet`s? Symex-driven loading will solve this properly, as stubs are produced at the same time as regular methods, so perhaps we shouldn't go too far down this rabbit hole at this point.

File: src/util/std_types.h [hunk 1]
```diff
@@ -151,6 +151,13 @@ inline symbol_typet &to_symbol_type(typet &type)
   return static_cast<symbol_typet &>(type);
 }
 
+template <>
+inline bool can_cast_type<symbol_typet>(const typet &type)
```
Comment (tautschnig) [inline src/util/std_types.h:155]: Could this please be factored out into a separate commit and PR? It's the only reason I get invited to the reviewing party, and this bit looks good and useful to me. For all the rest I have no clue...

Comment (thk123) [reply src/util/std_types.h:155]: Sorry I didn't think the faff of two branches, having to rebase the later on the former when one is merged etc was worth the hassle - feel free to disregard the review request 

FWIW I increasingly view review requests that are automatically generated from code ownership are spurious and I wait till someone explicitly requests me before looking.

**Comment by thk123** [comment]
@smowton The method was being loaded here:

```
lazy_goto_modelt::load_all_functions()
  goto_functions.ensure_function_loaded(symbol_name);
```

which I thought was symex driven lazy loading code?

I agree that a good solution would rely on stubs being generated before lazy methods, which I guess is non-trivial without essentially re-inventing symex driven lazy loading? 

Based on above comment will modify this to just load the return type as I've done and add tests to that effect.

**Comment by smowton** [comment]
You might want to do return + reachable types though, since e.g. a stub that returns an `A { B b; }` will also result in a reachable `B`.

**Comment by smowton** [comment]
`load_all_functions` is the lazy-goto-model framework operating in non-lazy mode (in this case, after ci-lazy has done its symbol-table trimming). I guess ci-lazy was retaining the *symbol*, without a body, then load-all was populating it -- I didn't realise ci-lazy and load-all would interact that way!

**Comment by thk123** [comment]
@smowton is there already an easy way to accomplish this? None of the other calls to `add_needed_class` seem to recurse over the children?

**Comment by smowton** [comment]
This is true, because most instantiation sites only make one instance. However, parameters to the entry point and referenced globals are handled this way -- see for example https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/ci_lazy_methods.cpp#L370. That method (`initialize_all_instantiated_classes_from_pointer`) might be exactly what you need.

**Comment by thk123** [comment]
@smowton done as suggested (please look at the last two commits) will now add some jbmc tests that exercise this new feature.

**Review by smowton — APPROVED** [review]
That looks great! Thanks very much for the improvement.

**Comment by thk123** [comment]
Test added (if you'd care to re-review @smowton) - now just waiting on the backlog of CBMC->TG PRs.

File: jbmc/regression/jbmc/lazyloading_opaquereturn/Bar.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+public class Bar {
+  public int x;
+
+  public void cproverNondetInitialize() {
+
+  }
+}
```
Comment (thk123) [inline jbmc/regression/jbmc/lazyloading_opaquereturn/Bar.java]: Missing new lines!

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 3]
```diff
@@ -1039,6 +1039,14 @@ bool java_bytecode_languaget::convert_single_method(
     return false;
   }
 
+  const code_typet function_type = to_code_type(symbol.type);
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: Stick a why here

**Comment by thk123** [comment]
@smowton Sorry for asking for another re-review, but CI was failing due to `removed_virtual_functions` which was failing due to the added code was forcibly instantiating abstract classes. I think this is _never_ right and made this change -https://github.com/diffblue/cbmc/pull/2335/commits/2079e254657ee72c195d51672025d91b764a660b

**Comment by thk123** [comment]
CI failure due to new test not being disabled for symex driven lazy loading (despite using --lazy-methods) making it incompatible - now disabled so hopefully CI will pass. Will rebase once #2302 is merged.

**Comment by smowton** [comment]
Hmmmmm. How was this working before when a parameter had an abstract or interface type? On the face of it, yes, we should never regard an abstract type as instantiated. On the other, we have previously used a pseudo-instance of an abstract type to stand in for "an unknown subclass of this type" -- for example, a stub method with return type `Set` would return an actual instance of it, which notionally meant it's some concrete subtype but we don't know which one. Obviously that's a pretty spurious way of dealing with the problem, but it's also a reason to be careful with this sort of change.

Could you elaborate a bit on what problem was happening when we thought (e.g.) `Set` could be instantiated, and why that only occurred here and not for input parameters with `Set` type?

**Comment by thk123** [comment]
@smowton Hmm you're right, I didn't intend to change that behaviour. I'd prefer not to dig into that _right now_ so I will change to just not instantiate opaque methods returning abstract classes as this fixes the immediate problem - would you be happy with that?

**Comment by thk123** [comment]
FWIW, it seems to make no difference:

```
abstract class AbstractClass {
  abstract void foo();
}

public class TestClass {
  static void testFunction(AbstractClass input) {
    input.foo();
  }
}
```

```
CI lazy methods: elaborate java::AbstractClass.foo:()V
```

**Comment by smowton** [comment]
Sure, if you're reducing the ambition of the PR to stay on the safe side that's fine with me

**Comment by thk123** [comment]
So doing this hear has the downside that if the stub uses the pointer type selector to load a non-abstract class, then we won't fix that case. However, I will leave that as a separate issue to tackle (I believe it is just a case of moving the check inside the `add_all_needed_classes` but due to the unknowns discussed about ~I will raise a ticket for it~. and I can't even demonstrate this problem so will assume I'm just missing something.

**Comment by thk123** [comment]
@smowton Done as I said: https://github.com/diffblue/cbmc/pull/2335/commits/96eb10b476249117df119426ae21ff025bf1b871

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 96eb10b476249117df119426ae21ff025bf1b871).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 96eb10b476249117df119426ae21ff025bf1b871).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 96eb10b476249117df119426ae21ff025bf1b871).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Comment by thk123** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/1958

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 771dddb79e181cfa8f3dc5bb88a58a59d2c8e124).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76555658

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 4]
```diff
@@ -1038,6 +1038,26 @@ bool java_bytecode_languaget::convert_single_method(
...
+      const symbol_typet *symbol_type =
+        type_try_dynamic_cast<symbol_typet>(return_type_pointer->subtype());
+      if(symbol_type)
+      {
+        needed_lazy_methods->add_needed_class(symbol_type->get_identifier());
+      }
+    }
+  };
+
+  // opaque method, but we should load the classes of the parameters and return??
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: Incomplete comment?

File: jbmc/regression/jbmc/lazyloading_opaquereturn/Opaque.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+public class Opaque
```
Comment (jeannielynnmoulton) [inline jbmc/regression/jbmc/lazyloading_opaquereturn/Opaque.java:1]: Is this folder missing the script that deletes the opaque class?

File: jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp [hunk 6]
```diff
@@ -41,3 +44,108 @@ bool ci_lazy_methods_neededt::add_needed_class(
 
   return true;
 }
+
+/// Add to the needed classes all class specified, the replacement type if it
```
Comment (jeannielynnmoulton) [inline jbmc/src/java_bytecode/ci_lazy_methods_needed.cpp]: "all classes specified" ~and "fields"~

**Review by jeannielynnmoulton — APPROVED** [review]

File: jbmc/regression/jbmc/lazyloading_opaquereturn/Opaque.java (hunk 5 above)
Comment (thk123) [reply jbmc/regression/jbmc/lazyloading_opaquereturn/Opaque.java:1]: No - the `Opaque.class` is not checked in unlike the other `.class` files, do you think including a build script is worthwhile? I think it is in general not done but that's isn't an argument against starting.

Comment (jeannielynnmoulton) [reply jbmc/regression/jbmc/lazyloading_opaquereturn/Opaque.java:1]: I missed that it was not included and was thinking of java-compile-tests.

File: jbmc/src/java_bytecode/java_bytecode_language.cpp (hunk 4 above)
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_language.cpp]: I'll squash some commits so this goes away.

**Comment by thk123** [comment]
Comments addressed - awaiting CI

**Comment by thk123** [comment]
CI failure looks transient (no output at all) re-running.

File: src/util/std_types.h (hunk 1 above)
Comment (thk123) [reply src/util/std_types.h:155]: Heh - just for the record I now want this in another PR, so I have been hoisted by my own laziness.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 7e5f136b1b973f0ed329f52cdae91c457abd8be2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76702499
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e95f59d86e2870e02df7f6ee8d9a20193941ec87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76928514
